ddtrace/_trace/processor/__init__.py

```diff
@@ -292,18 +292,32 @@ class _Trace(object):
         type=Dict[str, DefaultDict],
     )
 
-    def on_span_start(self, span):
-        # type: (Span) -> None
+    def _on_span_start(self, span: Span) -> None:
+        """This function MUST be called with self._lock held."""
```

Comment:
The strong wording of this note makes me wonder if there's a way to enforce the condition - something like `assert self._lock.held` (not sure if a suitable interface actually exists)

Reply:
We could do something like
```suggestion
        """This function MUST be called with self._lock held."""
        if not self._lock.locked():
            log.warning("_on_span_start() called without holding self._lock")
            return
```
(since some people don't like to `assert` in production code)

Reply:
Turned out to be a little uglier than I'd like because the `Lock` and `RLock` APIs differ.